File: src-tauri/src/app_prefs.rs
//! PJ-229 — **app-global preferences that survive a restart.**
//!
//! Stored at `{app_data_dir}/app-prefs.json`, a sibling of the universe registry and of
//! `style-presets.json` — deliberately NOT inside any universe.
//!
//! Its first tenant is the interface language. That lived only in the WebView's
//! `localStorage`, which this project has already proved non-durable (PJ-110, the leveldb
//! orphan-wipe), and on 2026-08-08 the Boss closed Constellation in Arabic and reopened
//! it in English. Everything else he sets is written to a real file; the language was the
//! one preference riding on browser storage.
//!
//! **Why not `.constellation/settings.json`:** that file is per-UNIVERSE, so the language
//! would change when he switched universes; it does not exist on the first-run screen,
//! which has no universe yet; and its save path is latched behind a successful read, so
//! on exactly that screen the write would be refused — reproducing PJ-229 through a
//! different door.
//!
//! Rust is a dumb JSON store here, as in `style_presets`: the frontend owns the shape.
//! The two disciplines that module paid for in blood are inherited deliberately —
//! **strict read** (a corrupt file errors instead of returning `{}` that the next save
//! would flatten) and **atomic write** (a partial file is not a corrupt one).

use std::fs;
use std::path::PathBuf;
use tauri::Manager;

fn prefs_path(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let app_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data dir: {}", e))?;
    fs::create_dir_all(&app_dir)
        .map_err(|e| format!("Failed to create app data dir: {}", e))?;
    Ok(app_dir.join("app-prefs.json"))
}

/// The whole preferences object. `{}` when the file has never been written — which is
/// every install before this shipped, and is not an error.
///
/// A file that EXISTS but cannot be read or parsed IS an error: returning `{}` there
/// would tell the frontend "you have no preferences", and the next save would write that
/// emptiness over preferences that are still on disk.
#[tauri::command]
pub fn load_app_prefs(app: tauri::AppHandle) -> Result<serde_json::Value, String> {
    let path = prefs_path(&app)?;
    if !path.exists() {
        return Ok(serde_json::json!({}));
    }
    Ok(crate::universe::read_persisted_json::<serde_json::Value>(&path)?
        .unwrap_or_else(|| serde_json::json!({})))
}

/// Persist the whole preferences object. The frontend merges; this only stores.
#[tauri::command]
pub fn save_app_prefs(app: tauri::AppHandle, prefs: serde_json::Value) -> Result<(), String> {
    let path = prefs_path(&app)?;
    let data = serde_json::to_string_pretty(&prefs).map_err(|e| e.to_string())?;
    crate::universe::atomic_write(&path, data.as_bytes())
        .map_err(|e| format!("Failed to save app preferences: {}", e))
}

File: src/lib/appPrefs.ts
/**
 * PJ-229 — app-global preferences that survive a restart.
 *
 * Backed by `{app_data_dir}/app-prefs.json` (`src-tauri/src/app_prefs.rs`), a sibling of
 * the universe registry — NOT inside any universe, because its first tenant is the
 * interface language and that must not change when the user switches universe.
 *
 * ## Why this exists
 *
 * The language lived only in `localStorage`, which PJ-110 proved this app can lose (the
 * leveldb orphan-wipe). On 2026-08-08 the Boss closed Constellation in Arabic and it
 * reopened in English. localStorage is now a CACHE — see `reconcileLocaleFromDisk` in
 * `$lib/i18n` — and this file is the record.
 *
 * ## The read-succeeded latch
 *
 * `save` refuses until a `load` has succeeded. Writing a merged object on top of a read
 * we could not perform would flatten preferences that are still on disk — the same
 * failure `style_presets` documents, and the same discipline `settingsLoaded` already
 * applies in `libraries/store.ts`. A session that cannot read simply does not persist.
 */
import { invoke } from '@tauri-apps/api/core';

type Prefs = Record<string, unknown>;

let snapshot: Prefs = {};
let loaded = false;

/** Read the whole object once at boot. Never throws — a failure means "do not persist". */
export async function loadAppPrefs(): Promise<Prefs> {
	try {
		const v = await invoke<Prefs>('load_app_prefs');
		snapshot = v && typeof v === 'object' ? v : {};
		loaded = true;
	} catch (e) {
		// Deliberately not `loaded = true`: see the latch note above.
		console.warn('[appPrefs] could not read app-prefs.json; this session will not persist', e);
	}
	return snapshot;
}

/**
 * Merge a patch into the snapshot and write the whole object.
 *
 * No debounce: these are rare, deliberate acts (choosing a language), and the 300 ms
 * debounce the per-universe settings save uses is a loss window there is no reason to
 * inherit for a once-in-a-while write.
 */
export async function saveAppPrefs(patch: Prefs): Promise<boolean> {
	if (!loaded) return false;
	snapshot = { ...snapshot, ...patch };
	try {
		await invoke('save_app_prefs', { prefs: snapshot });
		return true;
	} catch (e) {
		// PJ-207 §15 — the caller is TOLD. This used to end at a console.warn, which a release
		// build has no console for: the interface language changed on screen, the write failed,
		// and the next launch quietly came back in the old language — which is the exact
		// complaint (PJ-110, 2026-08-08) this file was written to fix.
		console.warn('[appPrefs] could not save app-prefs.json', e);
		return false;
	}
}

/** The value read at boot, for callers that need it before any save. */
export function getAppPref<T>(key: string): T | undefined {
	return snapshot[key] as T | undefined;
}
